### PR TITLE
Fix scheduled report description double encoding error and provide way to do proper sanitization bit by bit.

### LINKED ISSUE #4231: Fix inconsistent sanitization: sanitize on output rather than input
The way in which Piwik deals w/ sanitizing user data is inconsistent: it is sometimes done on **input** (data is stored sanitized), sometimes on **output** (less frequently). For example, website names are stored in the DB sanitized and need to be unsanitized or outputted raw in HTML. Goal names are not sanitized in the DB and need to be escaped when outputting in HTML.

This should be made consistent. Additionally, [security best practices recommend](http://blog.ircmaxell.com/2011/03/what-is-security-web-application.html):
- filter on input (e.g. cast to an integer if you expect an integer, but don't sanitize strings)
- escape on **output**

Problems with the current approach:
- escaping on input is done for HTML/XML: that escaping doesn't makes sense for SQL, JavaScript/JSON, … And it needs to be undone for those cases
- double escaping issues if we escape on output (e.g. Twig and Angular do that by default)
  - #8496, #8123, #7987, #7969, #7806, #7531, #7528, #6821, #6722, #6325, #6068, #5189, #5009, #4749, #4709, #4231, #3954, #3549, #3503, #2519, #2400, #2386, #974, #341
- strings are stored in database escaped (not as their "real" values)

We should try to slowly move to sanitizing on output using native escaping features of Twig or AngularJS for example.

### PR DESCRIPTION
**Pull request by diosmosis**
Fixes the double encoding error in #7987 by storing scheduled reports descriptions unsanitized and sanitizing only on outputting to HTML.

Also includes the following changes which will allow us to fix #4231 one change at a time:
- Created RequestSanitizer service and moved all sanitization logic to it (all Common::(un)sanitizeInputValue\* methods).
- Changed API\Request's request parameter accessing & sanitizing logic so it didn't use `Common::getRequestVar()` and instead just accessed the value manually. Moved the sanitization logic for this to a method in RequestSanitizer.
- Allow DI config to blacklist certain API method parameters from being sanitized. This allows us to change one part of Piwik at a time to input only sanitization. For example, we can blacklist the add/update site methods when making sure site names are stored unsanitized in the DB in one PR, then in another PR deal w/ another part of Piwik.

Eventually, after all the input sanitization is removed in favor of output sanitization, the RequestSanitizer class should be completely removed. Should be done for 3.0, since it will change the API.
### Extra Changes
- Fixed an issue w/ double encoding in exception messages. Exception messages were always escaped, even for non-XML formats (eg, JSON). I changed this so only XML/HTML formats escape messages. Includes tests.
- Sanitize HTML provided to notification.js & Piwik_Popover.showError() so there will be no XSS when showing exception messages (or other messages) that contain HTML. Uses angular.js `$sanitize` service.
### Sanitization Change Workflow

With this approach, you can change one part of Piwik to use the sanitize on output approach by doing the following:
1. Find the API method that creates/updates the entity in question. For the API method parameter that should be stored unsanitized, blacklist it through DI config. This will remove sanitization for that specific parameter.
   
   To blacklist an API parameter, add the method and parameter name to the `'RequestSanitizer.requestParameterSanitizeBlacklist'` array in DI config (see ScheduledReports/config/config.php for an example).
2. Make sure the property is stored w/o sanitization. Normally, you won't have to do anything in this regard.
3. Make sure the property is properly escaped when outputting HTML/XML.
4. Test all places where the property can be displayed in the UI for XSS. (including exception messages)

Once all parts of Piwik are changed, we can make the breaking change of not sanitizing API method parameters before calling the method in 3.0.

TODO:
- [x] add issue for removing magic quote hack/workaround in Piwik 3.0
- [x] create new Update in ScheduledReports that unsanitizes report descriptions in the DB change
- [x] make sure to test all other scheduled reports formats + XSS in scheduled reports pages

Refs #7987 and #4231

### CODE REVIEW

File: core/API/Proxy.php [hunk 1]
```diff
@@ -374,42 +381,28 @@ public function setHideIgnoredFunctions($hideIgnoredFunctions)
...
-                        }
-                    }
-                }
-            } catch (Exception $e) {
+            // copied from Common::getRequestVar. will use default value in case of empty string which appears to be necessary.
+            // (tests fail otherwise)
+            $isValuePresent = isset($parametersRequest[$name]) && (is_array($parametersRequest[$name]) || strlen($parametersRequest[$name]) !== 0);
+
+            $value = $isValuePresent ? $parametersRequest[$name] : $defaultValue;
+            if ($value instanceof NoDefaultValue) {
```
Comment (mattab) [inline core/API/Proxy.php]: maybe this logic + throwing exception could be refactored in a method used here and in `Common::getRequestVar`?

**Comment by mattab** [comment]
Feedback
 * the concept and solution looks good to me :+1: 
 * there are a few unit tests failures: https://travis-ci.org/piwik/piwik/jobs/64028866
  * also some integration test failures but these could be random?

File: core/API/Proxy.php (hunk 1 above)
Comment (diosmosis) [inline core/API/Proxy.php]: Well, it's just the first line that was copied from getRequestVar, but moving it makes a lot of sense. Would clear up part of getRequestVar, too. Already done locally ;)

File: core/Common.php [hunk 2]
```diff
@@ -464,13 +365,15 @@ public static function getRequestVar($varName, $varDefault = null, $varType = nu
             }
         }
 
+        // return isset($parametersRequest[$name]) && (is_array($parametersRequest[$name]) || strlen($parametersRequest[$name]) !== 0);
+
```
Comment (mnapoli) [inline core/Common.php]: commented code

File: core/Http/RequestSanitizer.php [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+     *
+     * @param string $value Parses a JSON string into an array and then sanitizes every element recursively. Array
+     *                      keys are sanitized as well.
+     * @return array
+     */
+    public function sanitizeJsonValues($value)
+    {
+        $value = $this->undoMagicQuotes($value);
+        $value = json_decode($value, $assoc = true);
+        return $this->sanitizeValues($value, $alreadyStripslashed = true);
```
Comment (mnapoli) [inline core/Http/RequestSanitizer.php]: Can the value be put through `addslashes()` instead? That would avoid all the `alreadyStripslashed` parameters which kind of "pollute" the public API of the class with implementation detail. Not that it's a huge deal but if it can be easily avoided?

File: plugins/ScheduledReports/Updates/2.14.0-b2.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Update for version 2.14.0-b2.
+ */
+class Updates_2_14_0_b2 extends Updates
+{
+    public function doUpdate(Updater $updater)
+    {
+        $reportTable = Common::prefixTable('report');
+        $updateSingleReportDescSql = "UPDATE " . $reportTable . " SET description = \"%s\" WHERE idreport = %s";
```
Comment (mnapoli) [inline plugins/ScheduledReports/Updates/2.14.0-b2.php]: Is it possible to use prepared statements instead?

File: core/Http/RequestSanitizer.php (hunk 3 above)
Comment (diosmosis) [inline core/Http/RequestSanitizer.php]: This is old code, has to do with magic quotes. Not looking to change this in this PR.

File: plugins/ScheduledReports/Updates/2.14.0-b2.php (hunk 4 above)
Comment (diosmosis) [inline plugins/ScheduledReports/Updates/2.14.0-b2.php]: Would have to change Updater.php, which uses exec. Could do that in another pr first.

File: core/Http/RequestSanitizer.php (hunk 3 above)
Comment (mnapoli) [inline core/Http/RequestSanitizer.php]: OK fair enough :)

**Comment by mnapoli** [comment]
Would it make sense to have `RequestSanitizer` be just a `Sanitizer`? After all it isn't coupled to the request, it only sanitizes values.

Regarding putting the "blacklist" in the DI config, it works, but it's a bit "disconnected" from the controllers (i.e. I work on a controller, it's not easy to see if request parameters are escaped or not, which is a little critical). What about using an annotation for that? e.g.

```php
class Controller
{
    /**
     * @Escape(false)
     */
    public function saveReport($name, $description, ...)
    ...
}
```

(could be `@NoEscape` or whatever else)

File: core/Http/RequestSanitizer.php [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    /**
+     * Sanitizes a single value.
+     *
+     * @param mixed $value If numeric, the value is not changed. If a string, line breaks are removed and HTML entities
+     *                     are escaped.
+     * @param bool $alreadyStripslashed Implementation detail, ignore.
+     * @return string
+     * @throws \Exception If the value is null or a boolean.
+     */
+    public function sanitizeValue($value, $alreadyStripslashed = false)
```
Comment (mnapoli) [inline core/Http/RequestSanitizer.php]: How about simpler names, `sanitize`?

File: core/Http/RequestSanitizer.php [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        return $value;
+    }
+
+    /**
+     * Unsanitizes a single value. Replaces escaped HTML entities with the actual characters.
+     *
+     * @param string $value
+     * @return string
+     */
+    public function unsanitizeValue($value)
```
Comment (mnapoli) [inline core/Http/RequestSanitizer.php]: unsanitize?

File: core/Http/RequestSanitizer.php [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+    /**
+     * Sanitize an array of values recursively.
+     *
+     * @param mixed $value If an array, it is sanitized recursively. Keys are sanitized as well as values. If not,
+     *                     it is sanitized normally.
+     * @param bool $alreadyStripslashed Implementation detail, ignore.
+     * @return mixed
+     */
+    public function sanitizeValues($value, $alreadyStripslashed = false)
```
Comment (mnapoli) [inline core/Http/RequestSanitizer.php]: This could be merged with `sanitize()` because it already accepts `mixed` so there's not really a reason to have 2 methods.

Comment (diosmosis) [inline core/Http/RequestSanitizer.php]: I don't think it can. sanitizeValues() (formerly Common::sanitizeInputValues) which is used by Common::getRequestVar and other places recurses, however sanitizeValue (formerly Common::sanitizeInputValue) does not. If we merge them, then callers of sanitizeInputValue will suddenly recurse. It may not cause an issue, but the intent is to remove this entire class in 3.0, so I don't think it's really necessary to refactor then test all of Piwik for regressions.

File: core/Http/RequestSanitizer.php (hunk 5 above)
Comment (diosmosis) [inline core/Http/RequestSanitizer.php]: I can make this change. Note: this class should be removed completely for 3.0, so the names shouldn't matter too much.

**Comment by diosmosis** [comment]
> Would it make sense to have RequestSanitizer be just a Sanitizer? After all it isn't coupled to the request, it only sanitizes values (except for sanitizing API parameters and the blacklist).

The point of this class (which may not be mentioned in the issue description or class docs, the PR is still a WIP) is to centralize the logic so we can remove it easily in 3.0. We can call it Sanitizer, it's not really important to me what the name is. I went w/ RequestSanitizer since it is only used to sanitize query parameters in an HTTP request.

> Regarding putting the "blacklist" in the DI config, it works, but it's a bit "disconnected" from the controllers (i.e. I work on a controller, it's not easy to see if request parameters are escaped or not, which is a little critical). What about using an annotation for that? e.g.

It is disconnected, and it is supposed to be disconnected. Using this class & setup, we can change one part of Piwik at a time to be "store unsanitized, sanitize on output" by adding the info to DI (maintaining existing behavior in other places and thus preventing unwanted xss vulnerabilities). Then, when we change the default handling of parameters to be unsanitized in 3.0, we can just remove the class and DI entries. If we use annotations or something else coupled with the plugins, we'd have to modify every API method and remove the useless annotation.

Also note, this only affects APIs, so not Controllers (who don't actually get called w/ request parameters).

File: core/Http/RequestSanitizer.php (hunk 7 above)
Comment (mnapoli) [inline core/Http/RequestSanitizer.php]: OK I see, it's fine then. FYI I am not very confident we will be able to fix this whole sanitization issue in 3.0 (and thus remove this class), I mean it's a huge-major thing with high risk of creating security issues (XSS) while refactoring… Somehow I think it won't be as fast as we would like to

**Comment by mnapoli** [comment]
OK I see your point (it's all only temporary). FYI I am not very confident we will be able to fix this whole sanitization issue in 3.0 (and thus remove this class), I mean it's a huge-major thing with high risk of creating security issues (XSS) while refactoring… Somehow I think it won't be as fast as we would like to. That's why I was pushing with spending more time making this more "durable", but if you are confident then I don't want to slow you down :)

**Comment by diosmosis** [comment]
The ultimate goals is to do no sanitization on query parameters in API requests. That's a HUGE breaking change, so If we can't get to that point by 3.0, it will probably never happen. To make the process faster, maybe I can create an automated XSS test. The UI tests do this already to some extent, but it's certainly not complete. And maybe I can come up w/ a set of patterns for JS coding to avoid XSS (for example, some code builds HTML directly w/ the result of AJAX queries, this is not necessary w/ angularjs).

For an XSS test, we could potentially add `<script>$('body').html('xss ?')</script>` to every varchar/char/text/blob/etc. column we can find in the DB. Or just add the data by hand, and test that we've covered every potential table column. This would be hardening of what is currently done in UITestFixture. This would test having untrusted data in Piwik's DB.

I'm not sure how to test having untrusted input returned from Piwik (either because it is in the DB, or due to a man in the middle attack). We could fake XHR responses... but triggering specific code paths is impossible w/ the jquery soup much of the frontend consists of.

**Comment by diosmosis** [comment]
Note to merger: This PR could potentially introduce XSS, so it would be a good idea to do manual UI tests before merging.

**Comment by mattab** [comment]
> For an XSS test, we could potentially add <script>$('body').html('xss ?')</script> to every varchar/char/text/blob/etc. column we can find in the DB.

Unfortunately automated tests for XSS is not realistic nor practical for now...  we will need to manually check every instance of variable in all templates, and html generators, and manually sanitise the output for the context (js, html attribute, etc.). it's tedious work but needs to be done. Don't think we can do it by 3.0.0 as it's already huge scope https://github.com/piwik/piwik/milestones/3.0.0  but we shall discuss (#4231 is tentatively in 3.0.0)
 
> Note to merger: This PR could potentially introduce XSS

I don't think we should rely on the reviewer to catch XSS but rather the original code author. You don't write in PR that you checked instances of the use of variables and check they are sanitised... did you, or is this PR really risky in this regard?

**Comment by diosmosis** [comment]
> You don't write in PR that you checked instances of the use of variables and check they are sanitised...

I checked where the description property was used and where exception messages were used. Perhaps I checked every possible place, perhaps I missed something. Perhaps someone could double check.

**Comment by mattab** [comment]
> I checked where the description property was used and where exception messages were used. Perhaps I checked every possible place, perhaps I missed something. Perhaps someone could double check.

Ok thanks for more details, good to know :+1:

**Comment by mattab** [comment]
@diosmosis as you suspected, because this PR is quite complex, it didn't get merged in time. I guess it be a possibility to fix all sanitisations problems at once in #4231

File: core/Common.php [hunk 8]
```diff
@@ -263,33 +261,9 @@ public static function mb_strtolower($string)
...
-            }
-        } elseif (!is_null($value)
-            && !is_bool($value)
-        ) {
-            throw new Exception("The value to escape has not a supported type. Value = " . var_export($value, true));
-        }
-        return $value;
+        /** @var RequestSanitizer $requestSanitizer */
+        $requestSanitizer = StaticContainer::get('Piwik\Http\RequestSanitizer');
+        return $requestSanitizer->sanitizeValues($value, $alreadyStripslashed);
```
Comment (tsteur) [inline core/Common.php]: Did you check this for performance stuff and profile it? Eg we sometimes call this during bulk request hundred thousands of times. I presume `StaticContainer::get` should add too much overhead but just to be sure

Comment (diosmosis) [inline core/Common.php]: Tracking 600 requests takes 10.5s for me on master and ~11.5s on this branch. According to xhprof, it's not the `StaticContainer::get` call that's causing a problem, but that `Common::getRequestVar` is called twice as many times. Not sure why... maybe a faulty rebase. Still investigating.

Comment (diosmosis) [inline core/Common.php]: Ok, I can't really fix this now. It's due to StaticContainer, but I can't fix it w/o removing StaticContainer. Which I can't do w/o removing static functions like getRequestVar. It's not possible w/o a large commit.

I suggest, before we work on 3.0, we should create some xhprof profiles for specific parts of Piwik that we don't want performance regressions in (like tracking, archiving, all websites/multisites). We post the PHP benchmark code + xhprof profiles online somewhere. Then while working on 3.0, we can re-run the benchmarks periodically and make sure that we're not regressing, or at least how to eventually fix the regression, before the 3.0 release. What do you think?

If you agree, could you list the places we shouldn't regress? You've done most of the performance tuning, I think.

**Comment by mattab** [comment]
Realistically I think this won't get merged. Instead we shall try to fix the issue at once everywhere, which should make things clearer and easier in the mid term.
